lib/Wikia/src/CircuitBreaker/Storage/APCUAdapter.php

```diff
@@ -0,0 +1,174 @@
+<?php
+
+
+namespace Wikia\CircuitBreaker\Storage;
+
+
+use Ackintosh\Ganesha;
+use Ackintosh\Ganesha\Configuration;
+use Ackintosh\Ganesha\Storage\AdapterInterface;
+use Ackintosh\Ganesha\Storage\Adapter\TumblingTimeWindowInterface;
+use Ackintosh\Ganesha\Exception\StorageException;
+
+/**
+ * Class APCUAdapter
+ * @package Wikia\CircuitBreaker\Storage
+ */
+class APCUAdapter implements AdapterInterface, TumblingTimeWindowInterface {
+
+	/**
+	 * @var Configuration
+	 */
+	private $configuration;
+
+	private function readFailure( $key ) {
+		return "APCu fetch failed for key: {$key}";
+	}
+
+	private function storeFailure( $key ) {
+		return "APCu failed storing value for key: {$key}";
+	}
+
+	/**
+	 * @param Configuration $configuration
+	 * @return void
+	 */
+	public function setConfiguration( Configuration $configuration ) {
+		$this->configuration = $configuration;
+	}
+
+	/**
+	 * @param string $service
+	 * @return int
+	 */
+	public function load( $service ) {
+		$status = false;
+		$value = (int)apcu_fetch( $service, $status );
+		if ( $status ) {
+			return $value;
+		}
+
+		$this->throwException($this->readFailure( $service ));
+	}
+
+	/**
+	 * @param string $service
+	 * @param int $count
+	 * @return void
+	 */
+	public function save( $service, $count ) {
+		if (apcu_store( $service, $count ) === false) {
+			$this->throwException( $this->storeFailure( $service ) );
+		}
+	}
+
+	/**
+	 * @param string $service
+	 * @return void
+	 */
+	public function increment( $service ) {
+		$status = false;
+		apcu_inc( $service, 1, $status );
+		if ( $status === false ) {
+			$this->throwException( $this->storeFailure( $service ) );
+		}
+	}
+
+	/**
+	 * decrement failure count
+	 *
+	 * If the operation would decrease the value below 0, the new value must be 0.
+	 *
+	 * @param string $service
+	 * @return void
+	 */
+	public function decrement( $service ) {
+		$status = false;
+		apcu_dec( $service, 1, $status );
+		if ( $status === false ) {
+			$this->throwException( $this->storeFailure( $service ) );
+		}
+	}
+
+	/**
+	 * sets last failure time
+	 *
+	 * @param string $service
+	 * @param int $lastFailureTime
+	 * @return void
+	 */
+	public function saveLastFailureTime( $service, $lastFailureTime ) {
+		//Interestingly, Ganesha seems to be using the same key for everything, or it is $service meaning different
```

Comment:
$service is not exactly the same in all Storage calls. It gets wrapped with prefix and suffix for each method type so there is no collision: https://github.com/ackintosh/ganesha/blob/df94cbabbe8e3ad892cbf1a98c672f8286d63ff4/src/Ganesha/Storage.php#L229